# For the following exercises, you'll need to load several datasets using the pydataset library. (If you get an error when trying to run the import below, use pip to install the pydataset package.)

- from pydataset import data

When the instructions say to load a dataset, you can pass the name of the dataset as a string to the data function to load the dataset. You can also view the documentation for the data set by passing the show_doc keyword argument.

mpg = data('mpg') 
- #load the dataset and store it in a variable
- #data('mpg', show_doc=True) 
- #view the documentation for the dataset


In [1]:
from pydataset import data
import pandas as pd
import numpy as np

# 1. Load the mpg dataset. Read the documentation for it, and use the data to answer these questions:

In [2]:
#creates a variable holding the mpg dataframe
mpg_df = data('mpg')
#to read the documentation of mpg
data('mpg', show_doc = True)

mpg

PyDataset Documentation (adopted from R Documentation. The displayed examples are in R)

## Fuel economy data from 1999 and 2008 for 38 popular models of car

### Description

This dataset contains a subset of the fuel economy data that the EPA makes
available on http://fueleconomy.gov. It contains only models which had a new
release every year between 1999 and 2008 - this was used as a proxy for the
popularity of the car.

### Usage

    data(mpg)

### Format

A data frame with 234 rows and 11 variables

### Details

  * manufacturer. 

  * model. 

  * displ. engine displacement, in litres 

  * year. 

  * cyl. number of cylinders 

  * trans. type of transmission 

  * drv. f = front-wheel drive, r = rear wheel drive, 4 = 4wd 

  * cty. city miles per gallon 

  * hwy. highway miles per gallon 

  * fl. 

  * class. 




### On average, which manufacturer has the best miles per gallon?

In [3]:
#creates a new column of the average of the city and highway mpg
mpg_df['overall_mpg'] = (mpg_df.cty + mpg_df.hwy) / 2
#creates a new df for the values grouped by the manufacturer
#displaying 1 shows us the highest overall mpg
average = mpg_df.groupby('manufacturer').mean()
average = average.sort_values(by='overall_mpg', ascending = False)
average.head(1)
#best_mpg = average.head(1)
#best_mpg

displ    year  cyl        cty        hwy  overall_mpg
manufacturer                                                          
honda         1.711111  2003.0  4.0  24.444444  32.555556         28.5

### How many different manufacturers are there?

In [4]:
#value counts finds the count of each unique value
#using count at the end, sums the unique manufactures
count = mpg_df.manufacturer.value_counts().count()
print('There are',count,'different maufactures in mpg.')

There are 15 different maufactures in mpg.


### How many different models are there?

In [5]:
#value counts finds the count of each unique value
#using count at the end, sums the unique models
count = mpg_df.model.value_counts().count()
print('There are',count,'different models in mpg.')

There are 38 different models in mpg.


### Do automatic or manual cars have better miles per gallon?

In [6]:
#creates a new df of only the automatic cars
auto = mpg_df[mpg_df.trans.str.contains('auto')]

#creates a new data frame of only the manual cars
manual = mpg_df[mpg_df.trans.str.contains('manual')]

#displays the average overall mpg of all automatic cars
print('The overall mpg of automatic cars is: ',auto.overall_mpg.mean())

#displays the average overall mpg of all manual cars
print('The overall mpg of manual cars is: ',manual.overall_mpg.mean())

#prints the higher mpg hardcoded
print('Manual cars have better overall mileage than automatic cars.')

The overall mpg of automatic cars is:  19.130573248407643
The overall mpg of manual cars is:  22.227272727272727
Manual cars have better overall mileage than automatic cars.


# 2. Joining and Merging

### Copy the users and roles dataframes from the examples above.

In [7]:
#creates users data frame
users = pd.DataFrame({
    'id': [1, 2, 3, 4, 5, 6],
    'name': ['bob', 'joe', 'sally', 'adam', 'jane', 'mike'],
    'role_id': [1, 2, 3, 3, np.nan, np.nan]
})

#creates roles data frame
roles = pd.DataFrame({
    'id': [1, 2, 3, 4],
    'name': ['admin', 'author', 'reviewer', 'commenter']
})

### A left join

In [8]:
#displaying a left join of users and roles
pd.merge(users, roles, left_on='role_id', right_on='id', how='left')

id_x name_x  role_id  id_y    name_y
0     1    bob      1.0   1.0     admin
1     2    joe      2.0   2.0    author
2     3  sally      3.0   3.0  reviewer
3     4   adam      3.0   3.0  reviewer
4     5   jane      NaN   NaN       NaN
5     6   mike      NaN   NaN       NaN

### What do you think a right join would look like?

We are considering the roles table as the right join. For a right join, it takes everything from the right table, and everything shared between the two tables. What is not shared in the left table with the right table is not included. So the values that are not defined in the left table are shown as NaN.
"Returns all records from the right table (table2), and the matched records from the left table (table1). The result is NULL from the left side, when there is no match."
https://www.w3schools.com/sql/sql_join_right.asp

In [9]:
#a right join of users and roles
pd.merge(users, roles, left_on='role_id', right_on='id', how='right')

id_x name_x  role_id  id_y     name_y
0   1.0    bob      1.0     1      admin
1   2.0    joe      2.0     2     author
2   3.0  sally      3.0     3   reviewer
3   4.0   adam      3.0     3   reviewer
4   NaN    NaN      NaN     4  commenter

### An outer join? 

In an outer join, unmatched rows in one or both tables can be returned.

In [10]:
#an outer join of users and roles
pd.merge(users, roles, left_on='role_id', right_on='id', how='outer')

id_x name_x  role_id  id_y     name_y
0   1.0    bob      1.0   1.0      admin
1   2.0    joe      2.0   2.0     author
2   3.0  sally      3.0   3.0   reviewer
3   4.0   adam      3.0   3.0   reviewer
4   5.0   jane      NaN   NaN        NaN
5   6.0   mike      NaN   NaN        NaN
6   NaN    NaN      NaN   4.0  commenter

### What happens if you drop the foreign keys from the dataframes and try to merge them?

In [11]:
#creating the tables without the shared key
#a join can not happen without these
#creates users data frame
users2 = pd.DataFrame({
    'id': [1, 2, 3, 4, 5, 6],
    'name': ['bob', 'joe', 'sally', 'adam', 'jane', 'mike'],
})

#creates roles data frame
roles2 = pd.DataFrame({
    'name': ['admin', 'author', 'reviewer', 'commenter']
})

In [12]:
pd.merge(users2, roles2, left_on='role_id', right_on='id', how='left')

KeyError: 'id'

# 3. Getting data from SQL databases

### Create a function named get_db_url. It should accept a username, hostname, password, and database name and return a url formatted like in the examples in this lesson.

In [13]:
from env import host, username, password

get_db_url = f'mysql+pymysql://{username}:{password}@{host}/employees'

### Use your function to obtain a connection to the employees database.

In [14]:
pd.read_sql('SELECT * FROM employees LIMIT 5 OFFSET 50', get_db_url)

emp_no  birth_date first_name  last_name gender   hire_date
0   10051  1953-07-28   Hidefumi      Caine      M  1992-10-15
1   10052  1961-02-26     Heping     Nitsch      M  1988-05-21
2   10053  1954-09-13     Sanjiv   Zschoche      F  1986-02-04
3   10054  1957-04-04     Mayumi  Schueller      M  1995-03-13
4   10055  1956-06-06     Georgy     Dredge      M  1992-04-27

### Once you have successfully run a query:

Intentionally make a typo in the database url. What kind of error message do you see?

In [15]:
get_db_url_typo = f'mysql+pymysql://{username}123:{password}@{host}/employees'

In [16]:
pd.read_sql('SELECT * FROM employees LIMIT 5 OFFSET 50', get_db_url_typo)

OperationalError: (pymysql.err.OperationalError) (1045, "Access denied for user 'darden_1031123'@'cpe-173-174-236-31.satx.res.rr.com' (using password: YES)")
(Background on this error at: http://sqlalche.me/e/e3q8)

Error message was...
OperationalError: (1045, "Access denied for user...

Intentionally make an error in your SQL query. What does the error message look like?

In [17]:
pd.read_sql('SELECT * FRO employees LIMIT 5 OFFSET 50', get_db_url)

ProgrammingError: (pymysql.err.ProgrammingError) (1064, "You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'FRO employees LIMIT 5 OFFSET 50' at line 1")
[SQL: SELECT * FRO employees LIMIT 5 OFFSET 50]
(Background on this error at: http://sqlalche.me/e/f405)

- when table spelled wrong: ProgrammingError: (1146, "Table 'employees.employee' doesn't exist")
- when syntax spelled wrong: ProgrammingError: (1064, "You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'FRO employees LIMIT 5 OFFSET 50' at line 1")

Read the employees and titles tables into two separate dataframes

In [18]:
employees = pd.read_sql('SELECT * FROM employees', get_db_url)
titles = pd.read_sql('SELECT * FROM titles', get_db_url)

Visualize the number of employees with each title.

In [19]:
#using historical data, we did not filter out the old title dates from above
#we joined the tables with a left join using the emp_no
emp_with_titles = pd.merge(employees, titles, left_on='emp_no', right_on='emp_no', how='left')
#value counts shows us the counts of the distinct title names
emp_with_titles.title.value_counts()

Engineer              115003
Staff                 107391
Senior Engineer        97750
Senior Staff           92853
Technique Leader       15159
Assistant Engineer     15128
Manager                   24
Name: title, dtype: int64

In [20]:
#this time we create a titles data table with only the current employee titles
titles_current = pd.read_sql('SELECT * FROM titles WHERE to_date > CURDATE()', get_db_url)
#we join the two together with the emp_no
emp_with_titles_current = pd.merge(employees, titles_current, left_on='emp_no', right_on='emp_no', how='left')
#value counts shows us the counts of the distinct title names
emp_with_titles_current.title.value_counts()

Senior Engineer       85939
Senior Staff          82024
Engineer              30983
Staff                 25526
Technique Leader      12055
Assistant Engineer     3588
Manager                   9
Name: title, dtype: int64

In [21]:
print(employees.head(5))
titles.head(5)

   emp_no  birth_date first_name last_name gender   hire_date
0   10001  1953-09-02     Georgi   Facello      M  1986-06-26
1   10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2   10003  1959-12-03      Parto   Bamford      M  1986-08-28
3   10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4   10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12


emp_no            title   from_date     to_date
0   10001  Senior Engineer  1986-06-26  9999-01-01
1   10002            Staff  1996-08-03  9999-01-01
2   10003  Senior Engineer  1995-12-03  9999-01-01
3   10004         Engineer  1986-12-01  1995-12-01
4   10004  Senior Engineer  1995-12-01  9999-01-01

Join the employees and titles dataframes together. Visualize how frequently employees change titles.

In [22]:
#we joined the two tables together as emp_with_titles
changes = emp_with_titles.emp_no.value_counts().sum()
emp_total = emp_with_titles.emp_no.value_counts().count()
title_rate = changes / emp_total
print('Employees change titles at a rate of', title_rate, 'per employee.')

Employees change titles at a rate of 1.4775751273231474 per employee.


For each title, find the hire date of the employee that was hired most recently with that title.

In [34]:
emp_with_titles.titles.value_counts()

AttributeError: 'DataFrame' object has no attribute 'titles'

Write the code necessary to create a cross tabulation of the number of titles by department. (Hint: this will involve a combination of SQL and python/pandas code)

# 4. Use your get_db_url function to help you explore the data from the chipotle database. Use the data to answer the following questions:

What is the total price for each order?

What are the most popular 3 items?

Which item has produced the most revenue?